state/src/state.rs

```diff
@@ -0,0 +1,658 @@
+            if !contain_key {
+                let trie = PatriciaTrie::from(&mut self.db, RLPNodeCodec::default(), &self.root.0)?;
+                match trie.get(&address)? {
+                    Some(rlp) => {
+                        let mut state_object = StateObject::from_rlp(&rlp)?;
+                        state_object.read_code(&mut self.db)?;
+                        state_object.set_storage(key, value);
+                        self.insert_cache(address, StateObjectEntry::new_dirty(Some(state_object)));
+                    }
+                    None => {
+                        return Err(Error::KeyNotFound);
+                    }
+                }
+            }
+        }
+        self.add_checkpoint(address);
+
+        if let Some(ref mut state_object_entry) = self.cache.borrow_mut().get_mut(address) {
+            match state_object_entry.state_object {
+                Some(ref mut state_object) => {
+                    state_object.set_storage(key, value);
+                    state_object_entry.status = ObjectStatus::Dirty;
+                }
+                None => panic!("state object always exist in cache."),
+            }
+        }
+        Ok(())
+    }
+
+    pub fn insert_cache(&self, address: &Address, state_object_entry: StateObjectEntry) {
+        let is_dirty = state_object_entry.is_dirty();
+        let old_entry = self
+            .cache
+            .borrow_mut()
+            .insert(*address, state_object_entry.clone_dirty());
+
+        if is_dirty {
+            if let Some(checkpoint) = self.checkpoints.borrow_mut().last_mut() {
+                checkpoint.entry(*address).or_insert(old_entry);
+            }
+        }
+    }
+
+    pub fn commit(&mut self) -> Result<(), Error> {
+        assert!(self.checkpoints.borrow().is_empty());
+
+        // firstly, update account storage tree
+        for (_address, entry) in self
+            .cache
+            .borrow_mut()
+            .iter_mut()
+            .filter(|&(_, ref a)| a.is_dirty())
+        {
+            if let Some(ref mut state_object) = entry.state_object {
+                state_object.commit_storage(&mut self.db)?;
+                state_object.commit_code(&mut self.db)?;
+            }
+        }
+
+        // secondly, update the whold state tree
+        let mut trie = PatriciaTrie::from(&mut self.db, RLPNodeCodec::default(), &self.root.0)?;
+        for (address, entry) in self
+            .cache
+            .borrow_mut()
+            .iter_mut()
+            .filter(|&(_, ref a)| a.is_dirty())
+        {
+            entry.status = ObjectStatus::Committed;
+            match entry.state_object {
+                Some(ref mut state_object) => {
+                    trie.insert(address, &rlp::encode(&state_object.account()))?;
+                }
+                None => {
+                    trie.remove(address)?;
+                }
+            }
+        }
+        self.root = From::from(&trie.root()?[..]);
+        Ok(())
```

Comment:
thirdly, we should update the cache?

Reply:
No. `Commit()` flush the data from cache to disk db, It's unnecessary to update the cache.
 Or your means is clear cache? A `clear()` function has been provided for this.